File: code/java/grpc/client/src/main/java/com/taskapi/grpc/client/TaskGrpcClient.java
package com.taskapi.grpc.client;

import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.StatusRuntimeException;
import io.grpc.stub.StreamObserver;
import lombok.extern.slf4j.Slf4j;
// Import generated classes (will be available after proto compilation)
// import com.taskapi.grpc.generated.*;

import java.util.*;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

@Slf4j
public class TaskGrpcClient {
    private final ManagedChannel channel;
    // private final TaskServiceGrpc.TaskServiceBlockingStub blockingStub;
    // private final TaskServiceGrpc.TaskServiceStub asyncStub;
    
    public TaskGrpcClient(String host, int port) {
        this(ManagedChannelBuilder.forAddress(host, port)
                .usePlaintext()
                .build());
    }
    
    public TaskGrpcClient(ManagedChannel channel) {
        this.channel = channel;
        // this.blockingStub = TaskServiceGrpc.newBlockingStub(channel);
        // this.asyncStub = TaskServiceGrpc.newStub(channel);
    }
    
    public void shutdown() throws InterruptedException {
        channel.shutdown().awaitTermination(5, TimeUnit.SECONDS);
    }
    
    /*
    public List<Task> listTasks(int pageSize, String pageToken, TaskStatus status,
                                 String assignedTo, List<String> tags, SortOrder sortOrder) {
        ListTasksRequest.Builder requestBuilder = ListTasksRequest.newBuilder()
                .setPageSize(pageSize);
        
        if (pageToken != null && !pageToken.isEmpty()) {
            requestBuilder.setPageToken(pageToken);
        }
        if (status != null) {
            requestBuilder.setStatus(status);
        }
        if (assignedTo != null && !assignedTo.isEmpty()) {
            requestBuilder.setAssignedTo(assignedTo);
        }
        if (tags != null && !tags.isEmpty()) {
            requestBuilder.addAllTags(tags);
        }
        if (sortOrder != null) {
            requestBuilder.setSortOrder(sortOrder);
        }
        
        List<Task> tasks = new ArrayList<>();
        try {
            Iterator<Task> response = blockingStub.listTasks(requestBuilder.build());
            while (response.hasNext()) {
                tasks.add(response.next());
            }
        } catch (StatusRuntimeException e) {
            log.error("RPC failed: {}", e.getStatus());
        }
        
        return tasks;
    }
    
    public Task getTask(String taskId) {
        GetTaskRequest request = GetTaskRequest.newBuilder()
                .setId(taskId)
                .build();
        
        try {
            return blockingStub.getTask(request);
        } catch (StatusRuntimeException e) {
            log.error("RPC failed: {}", e.getStatus());
            return null;
        }
    }
    
    public Task createTask(String title, String description, TaskPriority priority,
                           List<String> tags, String assignedTo) {
        Task.Builder taskBuilder = Task.newBuilder()
                .setTitle(title);
        
        if (description != null) {
            taskBuilder.setDescription(description);
        }
        if (priority != null) {
            taskBuilder.setPriority(priority);
        }
        if (tags != null) {
            taskBuilder.addAllTags(tags);
        }
        if (assignedTo != null) {
            taskBuilder.setAssignedTo(assignedTo);
        }
        
        CreateTaskRequest request = CreateTaskRequest.newBuilder()
                .setTask(taskBuilder.build())
                .build();
        
        try {
            return blockingStub.createTask(request);
        } catch (StatusRuntimeException e) {
            log.error("RPC failed: {}", e.getStatus());
            return null;
        }
    }
    
    public Task updateTask(String taskId, String title, String description,
                           TaskStatus status, TaskPriority priority,
                           List<String> tags, String assignedTo) {
        Task.Builder taskBuilder = Task.newBuilder()
                .setId(taskId);
        
        List<String> updateMask = new ArrayList<>();
        
        if (title != null) {
            taskBuilder.setTitle(title);
            updateMask.add("title");
        }
        if (description != null) {
            taskBuilder.setDescription(description);
            updateMask.add("description");
        }
        if (status != null) {
            taskBuilder.setStatus(status);
            updateMask.add("status");
        }
        if (priority != null) {
            taskBuilder.setPriority(priority);
            updateMask.add("priority");
        }
        if (tags != null) {
            taskBuilder.addAllTags(tags);
            updateMask.add("tags");
        }
        if (assignedTo != null) {
            taskBuilder.setAssignedTo(assignedTo);
            updateMask.add("assigned_to");
        }
        
        UpdateTaskRequest request = UpdateTaskRequest.newBuilder()
                .setTask(taskBuilder.build())
                .addAllUpdateMask(updateMask)
                .build();
        
        try {
            return blockingStub.updateTask(request);
        } catch (StatusRuntimeException e) {
            log.error("RPC failed: {}", e.getStatus());
            return null;
        }
    }
    
    public boolean deleteTask(String taskId) {
        DeleteTaskRequest request = DeleteTaskRequest.newBuilder()
                .setId(taskId)
                .build();
        
        try {
            blockingStub.deleteTask(request);
            return true;
        } catch (StatusRuntimeException e) {
            log.error("RPC failed: {}", e.getStatus());
            return false;
        }
    }
    
    public void watchTasks(boolean watchAll, List<String> taskIds, String assignedTo,
                           StreamObserver<TaskEvent> responseObserver) {
        StreamObserver<WatchTasksRequest> requestObserver = asyncStub.watchTasks(responseObserver);
        
        try {
            WatchTasksRequest.Builder requestBuilder = WatchTasksRequest.newBuilder()
                    .setWatchAll(watchAll);
            
            if (taskIds != null && !taskIds.isEmpty()) {
                requestBuilder.addAllTaskIds(taskIds);
            }
            if (assignedTo != null && !assignedTo.isEmpty()) {
                requestBuilder.setAssignedTo(assignedTo);
            }
            
            requestObserver.onNext(requestBuilder.build());
            
            // Keep the stream open for continuous updates
            // In a real application, you would manage this lifecycle appropriately
            
        } catch (Exception e) {
            log.error("Error in watch stream", e);
            requestObserver.onError(e);
        }
    }
    */
    
    public static void main(String[] args) throws InterruptedException {
        TaskGrpcClient client = new TaskGrpcClient("localhost", 50051);
        
        try {
            log.info("Note: This is a template. Run 'mvn compile' first to generate proto classes.");
            
            // Example usage (will work after proto generation):
            
            /*
            // Create a task
            log.info("Creating a new task...");
            Task newTask = client.createTask(
                    "Test Java gRPC Client",
                    "Testing the Java gRPC client implementation",
                    TaskPriority.HIGH,
                    Arrays.asList("test", "java", "grpc"),
                    "dev-team"
            );
            if (newTask != null) {
                log.info("Created task: {} - {}", newTask.getId(), newTask.getTitle());
                
                // Get the task
                log.info("\nRetrieving task {}...", newTask.getId());
                Task task = client.getTask(newTask.getId());
                if (task != null) {
                    log.info("Retrieved task: {} - Status: {}", task.getTitle(), task.getStatus());
                }
                
                // Update task status
                log.info("\nUpdating task status to IN_PROGRESS...");
                Task updated = client.updateTask(
                        newTask.getId(),
                        null,
                        null,
                        TaskStatus.IN_PROGRESS,
                        null,
                        null,
                        null
                );
                if (updated != null) {
                    log.info("Updated task status: {}", updated.getStatus());
                }
                
                // List all tasks
                log.info("\nListing all tasks...");
                List<Task> tasks = client.listTasks(10, null, null, null, null, SortOrder.CREATED_DESC);
                for (Task t : tasks) {
                    log.info("[{}] {} - {}", t.getStatus(), t.getTitle(), t.getId());
                }
                
                // Watch for task changes
                log.info("\nWatching for task changes...");
                CountDownLatch latch = new CountDownLatch(1);
                client.watchTasks(true, null, null, new StreamObserver<TaskEvent>() {
                    @Override
                    public void onNext(TaskEvent event) {
                        log.info("Event: {} - Task: {}", 
                                event.getEventType(), 
                                event.getTask().getTitle());
                    }
                    
                    @Override
                    public void onError(Throwable t) {
                        log.error("Watch error", t);
                        latch.countDown();
                    }
                    
                    @Override
                    public void onCompleted() {
                        log.info("Watch completed");
                        latch.countDown();
                    }
                });
                
                // Wait for a bit to receive events
                latch.await(5, TimeUnit.SECONDS);
                
                // Delete the task
                log.info("\nDeleting task {}...", newTask.getId());
                boolean deleted = client.deleteTask(newTask.getId());
                log.info("Task deleted: {}", deleted);
            }
            */
            
        } finally {
            client.shutdown();
        }
    }
}

File: code/kotlin/grpc/client/src/main/kotlin/com/taskapi/grpc/client/TaskGrpcClient.kt
package com.taskapi.grpc.client

import com.google.protobuf.util.Timestamps
import io.grpc.ManagedChannel
import io.grpc.ManagedChannelBuilder
import io.grpc.StatusException
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.toList
import java.util.concurrent.TimeUnit

// Note: These imports will be available after proto compilation
// import com.taskapi.grpc.*

class TaskGrpcClient(host: String, port: Int) {
    private val channel: ManagedChannel = ManagedChannelBuilder
        .forAddress(host, port)
        .usePlaintext()
        .build()
    
    private val stub = TaskServiceGrpcKt.TaskServiceCoroutineStub(channel)
    
    suspend fun listTasks(
        pageSize: Int = 20,
        pageToken: String = "",
        status: TaskStatus = TaskStatus.UNSPECIFIED,
        assignedTo: String = "",
        tags: List<String> = emptyList(),
        sortOrder: SortOrder = SortOrder.UNSPECIFIED
    ): List<Task> {
        val request = ListTasksRequest.newBuilder()
            .setPageSize(pageSize)
            .setPageToken(pageToken)
            .setStatus(status)
            .setAssignedTo(assignedTo)
            .addAllTags(tags)
            .setSortOrder(sortOrder)
            .build()
        
        return stub.listTasks(request).toList()
    }
    
    suspend fun getTask(taskId: String): Task? {
        return try {
            val request = GetTaskRequest.newBuilder()
                .setId(taskId)
                .build()
            
            stub.getTask(request)
        } catch (e: StatusException) {
            if (e.status.code == io.grpc.Status.Code.NOT_FOUND) {
                null
            } else {
                throw e
            }
        }
    }
    
    suspend fun createTask(
        title: String,
        description: String = "",
        priority: TaskPriority = TaskPriority.MEDIUM,
        tags: List<String> = emptyList(),
        assignedTo: String = ""
    ): Task {
        val task = Task.newBuilder()
            .setTitle(title)
            .setDescription(description)
            .setPriority(priority)
            .addAllTags(tags)
            .setAssignedTo(assignedTo)
            .build()
        
        val request = CreateTaskRequest.newBuilder()
            .setTask(task)
            .build()
        
        return stub.createTask(request)
    }
    
    suspend fun updateTask(
        taskId: String,
        title: String? = null,
        description: String? = null,
        status: TaskStatus? = null,
        priority: TaskPriority? = null,
        tags: List<String>? = null,
        assignedTo: String? = null
    ): Task? {
        return try {
            val taskBuilder = Task.newBuilder().setId(taskId)
            val updateMask = mutableListOf<String>()
            
            title?.let {
                taskBuilder.title = it
                updateMask.add("title")
            }
            description?.let {
                taskBuilder.description = it
                updateMask.add("description")
            }
            status?.let {
                taskBuilder.status = it
                updateMask.add("status")
            }
            priority?.let {
                taskBuilder.priority = it
                updateMask.add("priority")
            }
            tags?.let {
                taskBuilder.addAllTags(it)
                updateMask.add("tags")
            }
            assignedTo?.let {
                taskBuilder.assignedTo = it
                updateMask.add("assigned_to")
            }
            
            val request = UpdateTaskRequest.newBuilder()
                .setTask(taskBuilder.build())
                .addAllUpdateMask(updateMask)
                .build()
            
            stub.updateTask(request)
        } catch (e: StatusException) {
            if (e.status.code == io.grpc.Status.Code.NOT_FOUND) {
                null
            } else {
                throw e
            }
        }
    }
    
    suspend fun deleteTask(taskId: String): Boolean {
        return try {
            val request = DeleteTaskRequest.newBuilder()
                .setId(taskId)
                .build()
            
            stub.deleteTask(request)
            true
        } catch (e: StatusException) {
            if (e.status.code == io.grpc.Status.Code.NOT_FOUND) {
                false
            } else {
                throw e
            }
        }
    }
    
    suspend fun watchTasks(
        watchAll: Boolean = false,
        taskIds: List<String> = emptyList(),
        assignedTo: String = ""
    ): Flow<TaskEvent> {
        val requestFlow = flow {
            emit(
                WatchTasksRequest.newBuilder()
                    .setWatchAll(watchAll)
                    .addAllTaskIds(taskIds)
                    .setAssignedTo(assignedTo)
                    .build()
            )
        }
        
        return stub.watchTasks(requestFlow)
    }
    
    fun shutdown() {
        channel.shutdown().awaitTermination(5, TimeUnit.SECONDS)
    }
}

suspend fun main() {
    val client = TaskGrpcClient("localhost", 50051)
    
    try {
        println("Note: Run './gradlew build' first to generate proto classes.")
        
        // Example usage (will work after proto generation):
        
        /*
        // Create a task
        println("Creating a new task...")
        val task = client.createTask(
            title = "Test Kotlin gRPC Client",
            description = "Testing the Kotlin gRPC client implementation",
            priority = TaskPriority.HIGH,
            tags = listOf("test", "kotlin", "grpc"),
            assignedTo = "dev-team"
        )
        println("Created task: ${task.id} - ${task.title}")
        
        // Get the task
        println("\nRetrieving task ${task.id}...")
        val retrieved = client.getTask(task.id)
        if (retrieved != null) {
            println("Retrieved task: ${retrieved.title} - Status: ${retrieved.status}")
        }
        
        // Update task status
        println("\nUpdating task status to IN_PROGRESS...")
        val updated = client.updateTask(
            taskId = task.id,
            status = TaskStatus.IN_PROGRESS
        )
        if (updated != null) {
            println("Updated task status: ${updated.status}")
        }
        
        // List all tasks
        println("\nListing all tasks...")
        val tasks = client.listTasks(
            pageSize = 10,
            sortOrder = SortOrder.CREATED_DESC
        )
        tasks.forEach { t ->
            println("[${t.status}] ${t.title} - ${t.id}")
        }
        
        // Watch for task changes
        println("\nWatching for task changes...")
        var eventCount = 0
        client.watchTasks(watchAll = true).collect { event ->
            println("Event: ${event.eventType} - Task: ${event.task.title}")
            eventCount++
            if (eventCount >= 3) {
                return@collect
            }
        }
        
        // Delete the task
        println("\nDeleting task ${task.id}...")
        val deleted = client.deleteTask(task.id)
        println("Task deleted: $deleted")
        */
        
    } catch (e: Exception) {
        println("Error: ${e.message}")
        e.printStackTrace()
    } finally {
        client.shutdown()
    }
}